File: api/src/org/ufomatics/archipelago/model/AbstractDatabaseAdapter.java
package org.ufomatics.archipelago.model;

import java.util.Properties;

/**
 * @author <a href="mailto:rr0@rr0.org">Jerome Beau</a>
 * @version $revision$
 */
public abstract class AbstractDatabaseAdapter implements DatabaseAdapter {
    private Properties properties;

    protected AbstractDatabaseAdapter() {
    }

    protected abstract void init(Properties setupProperties) throws MetaException;

    public Properties getProperties() {
        return properties;
    }

    public void setProperties(Properties properties) {
        this.properties = properties;
        try {
            init (properties);
        } catch (MetaException e) {
            e.printStackTrace();
        }
    }

    public MetaObject visit(MetaObject original) throws MetaException {
        MetaObject readObject = new MetaObjectImpl(original.getType());
        readObject.setValues(original.getValues(), this);
        read(readObject);
        return readObject;
    }

    public abstract void read(MetaObject readObject) throws MetaException;
}

File: src/archipelago/api/model/AbstractDatabaseAdapter.ts
import {DatabaseAdapter} from "archipelago/api/model/DatabaseAdapter"
import {MetaObject} from "archipelago/api/model/MetaObject"
import {MetaObjectImpl} from "archipelago/api/model/MetaObjectImpl"
import {MetaModel} from "archipelago/api/model/MetaModel"
import {MetaModelImpl} from "archipelago/api/model/MetaModelImpl"
import {Properties} from "ts-jsdk"

/**
 *
 */
export abstract class AbstractDatabaseAdapter implements DatabaseAdapter {

  private properties: Properties = new Properties()

  constructor(readonly dataModel: MetaModel = new MetaModelImpl()) {
  }

  abstract getName(): string

  close(): void {
  }

  getProperties(): Properties {
    return this.properties
  }

  setProperties(properties: Properties): void {
    this.properties = properties
    try {
      this.init(properties)
    } catch (e) {
      console.error(e)
    }
  }

  visit(original: MetaObject): MetaObject {
    const readObject = new MetaObjectImpl(original.getType())
    readObject.setValues(original.getValues(), this)
    this.read(readObject)
    return readObject
  }

  abstract read(readObject: MetaObject): void

  protected abstract init(setupProperties: Properties): void
}
